model/riscv_step.sail

```diff
@@ -24,80 +136,111 @@ function step(step_no : int) -> bool = {
-          F_Error(e, addr) => {
-            handle_mem_exception(Virtaddr(addr), e);
-            (RETIRE_FAIL, false)
-          },
-          /* non-error cases: */
-          F_RVC(h) => {
-            sail_instr_announce(h);
-            instbits = zero_extend(h);
-            let ast = ext_decode_compressed(h);
-            if   get_config_print_instr()
-            then {
-              print_instr("[" ^ dec_str(step_no) ^ "] [" ^ to_str(cur_privilege) ^ "]: " ^ BitStr(PC) ^ " (" ^ BitStr(h) ^ ") " ^ to_str(ast));
-            };
-            /* check for RVC once here instead of every RVC execute clause. */
-            if currentlyEnabled(Ext_Zca) then {
-              nextPC = PC + 2;
-              (execute(ast), true)
-            } else {
-              handle_illegal();
-              (RETIRE_FAIL, true)
-            }
-          },
-          F_Base(w) => {
-            sail_instr_announce(w);
-            instbits = zero_extend(w);
-            let ast = ext_decode(w);
-            if   get_config_print_instr()
-            then {
-              print_instr("[" ^ dec_str(step_no) ^ "] [" ^ to_str(cur_privilege) ^ "]: " ^ BitStr(PC) ^ " (" ^ BitStr(w) ^ ") " ^ to_str(ast));
-            };
-            nextPC = PC + 4;
-            (execute(ast), true)
-          }
-        }
-      }
+  let ((step_val, did_step), instbits) : ((Step, bool), option(xlenbits)) =
+    match step_state {
+      STEP_WAIT   => (step_wait_state(step_no, exit_wait), None()),
+      STEP_ACTIVE => step_active_state(step_no, exit_wait)
     };
 
-  tick_pc();
+  var stepped : bool = did_step;
 
-  // Only increment minstret if the instruction retired successfully.
-  if retired != RETIRE_SUCCESS then minstret_increment = false;
-
-  update_minstret();
+  match step_val {
+    Step_Pending_Interrupt(intr, priv) => {
+      if   get_config_print_instr()
+      then print_bits("Handling interrupt: ", interruptType_to_bits(intr));
+      minstret_increment = false;
+      handle_interrupt(intr, priv)
+    },
+    Step_Ext_Fetch_Failure(e) => {
+      minstret_increment = false;
+      ext_handle_fetch_check_error(e)
+    },
+    Step_Waiting() => {
+      assert(~(wfi_is_nop()));
+      assert(step_state == STEP_WAIT, "cannot be Waiting in a non-Wait state");
+    },
+    Step_Execute(RETIRE_OK()) => {
+      assert(step_state == STEP_ACTIVE);
+      // default handling of minstret
+      ()
+    },
+    // standard errors
+    Step_Execute(RETIRE_FAIL(Trap(priv, ctl, pc))) => {
+      minstret_increment = false;
+      set_next_pc(exception_handler(priv, ctl, pc))
+    },
+    Step_Execute(RETIRE_FAIL(Memory_Exception(vaddr, exc))) => {
+      minstret_increment = false;
+      handle_mem_exception(vaddr, exc)
+    },
+    Step_Execute(RETIRE_FAIL(Illegal_Instruction())) => {
+      minstret_increment = false;
+      match instbits {
+        None()         => internal_error(__FILE__, __LINE__, "no instruction bits for illegal instruction"),
```

Comment:
This suggests the types are not quite right. Shouldn't `instbits` go in `Step_Execute`?

Reply:
It doesn't quite work; the `internal_error` just moves elsewhere.  The only way to avoid this is to give the `instbits` to `execute` so it can package it into `Illegal_Instruction`.  Or have it available in a global register, which is just going back to #412.  We can't package the instruction ast into `Illegal_Instruction` since it doesn't work for RVC instructions whose execute just calls the execute of the uncompressed ast.

Reply:
I was able to address this by storing the `instbits` in the `hart_state`.